tools/swift-inspect/Sources/AndroidCLib/heap.c

```diff
@@ -0,0 +1,114 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include <string.h>
+
+#include "heap.h"
+
+/* The heap metadata buffer is interpreted as an array of 8-byte pairs. The
+ * first pair contains metadata describing the buffer itself: max valid index
+ * (e.g. size of the buffer) and next index (e.g. write cursor/position). Each
+ * subsequent pair describes the address and length of a heap entry in the
+ * remote process. A 4KiB page provides sufficient space for the header and
+ * 255 (address, length) pairs.
+ *
+ * ------------
+ * | uint64_t | max valid index (e.g. sizeof(buffer) / sizeof(uint64_t))
+ * ------------
+ * | uint64_t | next free index (starts at 2)
+ * ------------
+ * | uint64_t | heap item 1 address
+ * ------------
+ * | uint64_t | heap item 1 size
+ * ------------
+ * | uint64_t | heap item 2 address
+ * ------------
+ * | uint64_t | heap item 2 size
+ * ------------
+ * | uint64_t | ...
+ * ------------
+ * | uint64_t | ...
+ * ------------
+ * | uint64_t | heap item N address
+ * ------------
+ * | uint64_t | heap item N size
+ * ------------
+ */
+
+#define MAX_VALID_IDX 0
+#define NEXT_FREE_IDX 1
+#define HEADER_SIZE 2
+#define ENTRY_SIZE  2
+
+#if defined(__aarch64__) || defined(__ARM64__) || defined(_M_ARM64)
+#define DEBUG_BREAK() asm("brk #0x0")
+#elif defined(_M_X64) || defined(__amd64__) || defined(__x86_64__) || defined(_M_AMD64)
+#define DEBUG_BREAK() asm("int3; nop")
+#else
+#error("only aarch64 and x86_64 are supported")
+#endif
```

Comment:
Ah, I was missing `__builtin_debugtrap()`-- that's the correct thing to do here. Since this is Android-only, it will always be clang and never msvc so we don't need the `__debugbreak` variant.